File: bit Operation位计算/78.subsets/Solution.py
def subsets2(nums):
    res = []
    nums.sort()
    for i in range(1<<len(nums)):
        tmp = []
        for j in range(len(nums)):
            temp1 = 1 << j
            temp2 = i & temp1
            if i & 1 << j:  # if i >> j & 1:
                tmp.append(nums[j])
        res.append(tmp)
    return res

if __name__ == '__main__':
    nums = [1,2,3]
    print(subsets2(nums))

"""
对于每一个subset，用其 下标（0，1，2，3，4，5，6，7）
分别与每个元素的下标&1（即 1， 2， 4）
进行
&运算

"""

File: bit Operation位计算/78.subsets/Solution.java
import java.util.ArrayList;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

// wrong answer, in fixing
public class Solution{
    public static List<List<Integer>> subsets(List<Integer> s){
        // Arrays.sort(s);
        // int ele_num = s.size();
        // int subset_sum = Math.pow(2, ele_num);
        // List<List<Integer>> subsets = new ArrayList<>();
        // for(int i = 0; i < ele_num; i++){
        //     List<Integer> subset = new ArrayList<>();
        //     for(int j = 0;j < subset_sum; j++){
        //         int temp1 = j >> i;
        //         boolean temp2 = temp1 & 1;
        //         if(temp2){
        //             subset.add(s.get(i));
        //         }
        //     }
        //     subsets.add(subset);
        // }
        // return subsets;
    }

    public static void main(String[] args) {
        List<Integer> s = List.of(1,2,3);
        List<List<Integer>> sets = subsets(s);
        System.out.println(sets);
    }

    // public static void printMatrix(List<List<Integer>> list){
    //     System.out.println("[");
    //     for(int i = 0; i < list.size(); i++){
    //         printMatrix(list.get(i) + ",");
    //     }
    //     System.out.println("]");
    // }
}